openssl3: Never use `__atomic_*` on macOS 10.7 and 10.8

catap opened the pull request:
#### Description

It also contains a rebase for existed patch for atomics.

Closes: https://trac.macports.org/ticket/64947


###### Type(s)
<!-- update (title contains ": U(u)pdate to"), submission (new Portfile) and CVE Identifiers are auto-detected, replace [ ] with [x] to select -->

- [x] bugfix
- [ ] enhancement
- [ ] security fix

###### Tested on
<!-- Triple-click and copy the next line and paste it into your shell. It will copy your OS and Xcode version to the clipboard. Paste it here replacing this section.
sh -c 'printf "%s\n" "macOS `sw_vers -productVersion` `sw_vers -buildVersion` `uname -m`" "`xcodebuild -version|awk '\''NR==1{x=$0}END{print x" "$NF}'\''`"'|tee /dev/tty|pbcopy
-->

macOS 10.7.5 11G63 x86_64
Xcode 4.6.3 4H1503

macOS 10.8.5 12F2560 x86_64
Xcode 5.1.1 5B1008

###### Verification <!-- (delete not applicable items) -->
Have you

- [x] followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?
- [x] squashed and [minimized your commits](https://guide.macports.org/#project.github)?
- [x] checked that there aren't other open [pull requests](https://github.com/macports/macports-ports/pulls) for the same change?
- [ ] referenced existing tickets on [Trac](https://trac.macports.org/wiki/Tickets) with full URL? <!-- Please don't open a new Trac ticket if you are submitting a pull request. -->
- [ ] checked your Portfile with `port lint --nitpick`?
- [ ] tried existing tests with `sudo port test`?
- [x] tried a full install with `sudo port -vst install`?
- [x] tested basic functionality of all binary files?

<!-- Use "skip notification" (surrounded with []) to avoid notifying maintainers -->

catap commented:
BTW I haven't got access to i386 with macOS and if anyone who had such access can try it, I'll be very appreciated.

kencu commented:
openssl3 already installs just fine for me on 10.7 without any modification:
```
$ uname -a
Darwin macbook21.local 11.4.2 Darwin Kernel Version 11.4.2: Thu Aug 23 16:26:45 PDT 2012; root:xnu-1699.32.7~1/RELEASE_I386 i386

$ port -v installed openssl3
The following ports are currently installed:
  openssl3 @3.0.2_0+legacy (active) requested_variants='' platform='darwin 11' archs='x86_64' date='2022-03-26T08:22:11-0700'
```

this looks perhaps to be a local issue on your system then?

Why don't you open a ticket and we'll see what is wrong with your openssl3 build there?

(I haven't tried 10.8).

kencu commented:
Looks like you are not using the right Xcode on 10.7 (and probably 10.8)
```
$ sh -c 'printf "%s\n" "macOS `sw_vers -productVersion` `sw_vers -buildVersion` `uname -m`" "`xcodebuild -version|awk '\''NR==1{x=$0}END{print x" "$NF}'\''`"'|tee /dev/tty|pbcopy
macOS 10.7.5 11G63 i386
Xcode 4.6.3 4H1503

$ clang -dM -E -x c /dev/null | grep apple
#define __apple_build_version__ 4250028
```

upgrade your Xcode and you should be fine (fine for me)

kencu commented:
As a PS -- I have sometimes seen workarounds to allow people to use less-than-the-last Xcode for older systems applied to some ports. I think this is very wrong-headed.

Supporting older systems is plenty hard enough, thank you. Supporting less-than-the-last Xcode on older systems is a crazy recipe for disaster, and we should absolutely not head down that rabbit hole.

kencu commented:
So what I missed here is that @catap is building using the libstdc++ standard library, which is probably why the buildbot and I see different results. I didn't realize this distinction at the beginning here, perhaps I missed reading where that was mentioned initially.

catap commented:
@Gcenx as I see you've used https://github.com/openssl/openssl/pull/16584 at https://github.com/macports/macports-ports/pull/12848 to fix some issues on i386. I assume that you still has access to such machine.

My PR is conflicted with old patch and I've rebased it (and simplified). May I ask your help to test it? Thanks!

Thus, this patch will be sent to upstream but after the first one is merged.

kencu commented:
Nope, it is not because of libstdc++:

```
$ cat /opt/bootstrap/etc/macports/macports.conf | grep stdlib
cxx_stdlib             libstdc++
```

```
$ port -v installed openssl3
The following ports are currently installed:
  openssl3 @3.0.2_0+legacy (active) requested_variants='' platform='darwin 11' archs='x86_64' date='2022-03-25T22:52:51-0700'
```

I continue to have absolutely no problem building openssl3 on 10.7 using the existing MacPorts portfiles, with no changes to anything, using either libstdc++ or libc++.

So the issue **must be** something specific to @catap 's 10.7 system therefore, and the only thing that we know for sure is that his command line tools have been shown to be out of date. As far as I know, his command line tools are still out of date, as he hasn't mentioned updating them. 

Please update your command line tools first, and see if that fixes this!

```
$ /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang -v
Apple LLVM version 4.2 (clang-425.0.28) (based on LLVM 3.2svn)
Target: x86_64-apple-darwin11.4.2
Thread model: posix


$ /usr/bin/clang -v
Apple LLVM version 4.2 (clang-425.0.28) (based on LLVM 3.2svn)
Target: x86_64-apple-darwin11.4.2
Thread model: posix
```

catap commented:
@kencu wired things:
```
macOS-Lion:~ catap$ /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang -v
Apple LLVM version 4.2 (clang-425.0.28) (based on LLVM 3.2svn)
Target: x86_64-apple-darwin11.4.2
Thread model: posix
macOS-Lion:~ catap$ /usr/bin/clang -v
Apple clang version 4.1 (tags/Apple/clang-421.11.66) (based on LLVM 3.1svn)
Target: x86_64-apple-darwin11.4.2
Thread model: posix
macOS-Lion:~ catap$ 
```
and `softwareupdate` says that everything is installed
```
macOS-Lion:~ catap$ sudo softwareupdate -l
Software Update Tool
Copyright 2002-2010 Apple

No new software available.
macOS-Lion:~ catap$
```
and it uses the right path:
```
macOS-Lion:~ catap$ xcode-select -print-path
/Applications/Xcode.app/Contents/Developer
macOS-Lion:~ catap$ 
```

any idea how can I update it?

catap commented:
@kencu I've also used this ansible playbook to install everythin on this machine: https://gist.github.com/catap/f9a4317302c1c16b74bee452da6122c6

It fails on `{{ bootstrap_prefix }}/bin/port -N install curl`.

Gcenx commented:
@catap added this PR into my local overlay and ran `port destroot openssl3 +universal`
```
Deans-MBP:~ gcenx$ file /opt/local/var/macports/build/_Users_gcenx_ports_devel_openssl3/openssl3/work/destroot/opt/local/libexec/openssl3/lib/libssl.3.dylib 
/opt/local/var/macports/build/_Users_gcenx_ports_devel_openssl3/openssl3/work/destroot/opt/local/libexec/openssl3/lib/libssl.3.dylib: Mach-O universal binary with 2 architectures: [i386:Mach-O dynamically linked shared library i386] [x86_64:Mach-O 64-bit dynamically linked shared library x86_64]
/opt/local/var/macports/build/_Users_gcenx_ports_devel_openssl3/openssl3/work/destroot/opt/local/libexec/openssl3/lib/libssl.3.dylib (for architecture i386):	Mach-O dynamically linked shared library i386
/opt/local/var/macports/build/_Users_gcenx_ports_devel_openssl3/openssl3/work/destroot/opt/local/libexec/openssl3/lib/libssl.3.dylib (for architecture x86_64):Mach-O 64-bit dynamically linked shared library x86_64
```

So at least the changes didn't seem to break anything on macOS Mojave, I don't have time to test on my 10.9 external install at the moment.

kencu commented:
So @catap -- as you can see, your command line tools are still out of date.

```
macOS-Lion:~ catap$ /usr/bin/clang -v
Apple clang version 4.1 (tags/Apple/clang-421.11.66) (based on LLVM 3.1svn)
Target: x86_64-apple-darwin11.4.2
Thread model: posix
macOS-Lion:~ catap$ 
```

please update them. I don't care what Software Update says. You need to get the right ones from the Apple Developer website and install them please. Then try your build of openssl3 again (and presumably watch it succeed, like everyone else's build does).

Otherwise these PRs are just totally wasting our time!

kencu commented:
I think this is the last one available on the apple website for Lion -- April 2013:

https://download.developer.apple.com/Developer_Tools/command_line_tools_os_x_lion_for_xcode__april_2013/xcode462_cltools_10_76938260a.dmg

You're looking for the one that gives you this:
```
$ /usr/bin/clang -v
Apple LLVM version 4.2 (clang-425.0.28) (based on LLVM 3.2svn)
Target: x86_64-apple-darwin11.4.2
Thread model: posix
```

kencu commented:
Once you get that installed, you should be able to install openssl3 on 10.7.5 as it is now -- works for me, no patches beyond what is already in the port.

catap commented:
@kencu thanks, you are right.

The root cause of issue that I've used too old command line tools. Why? Because I've download the upper one on apple developer web-site with assumption that it sorts them. I was wrong.

I'm going to close all my PRs.

Thanks.

kencu commented:
I'm so glad you have it sorted! Your input is extremely valuable, and I (we) want you in tip-top shape.

To be honest -- MacPorts port diagnose might have the smarts to do this. 

It is an extremely common problem that a user, even an extremely knowledgable user like you, has a mismatch between the Xcode and CLTs that are installed. I have had to track this down many times with users over the years, and in fact it has happened to me, so I well know how it makes tracking down inexplicable build errors hard.

It would not be too difficult perhaps to have a bit added to port diagnose that checks the CLTs and Xcode exact versions, compares them to a list of the latest ones for each system, and offers up at least the suggestion and maybe even better the download link for the right ones.

Gcenx commented:
I'd has a similar problem when playing with getting wine working on 10.6.8 so don't feel too bad about this.

catap commented:
@Gcenx I think that the right place is to decide at upstream of OpenSSL should it be merged or not.

kencu commented:
you don't want to needlessly disable good stuff like atomic support if you don't need to, though. Just makes things overall worse for no reason...
